src/lexer.coffee

```diff
@@ -1252,7 +1252,15 @@ REGEX = /// ^
 REGEX_FLAGS  = /^\w*/
 VALID_FLAGS  = /^(?!.*(.).*\1)[imguy]*$/
 
-HEREGEX      = /// ^(?: [^\\/#] | \\[\s\S] | /(?!//) | \#(?!\{) )* ///
+HEREGEX      = /// ^
+      (?:
+        [^\\/#]   |
+        \\[\s\S]  |
+        /(?!//)   |
+        # The comment should consume everything until the end of the line, including the '///'
+        \#(?!\{)(?:[^\n]*[\n])
```

Comment:
Sorry, one last little thing: The non-capturing group doesn’t do anything here, does it?

Reply:
I think it doesn't. But I found another bug which I missed before. 
Capturing comment is too aggressive, and following valid RegExp doesn't compile.
```coffeescript 
/// ^(?: [^\\"#] | \\[\s\S] |           \#(?)* ///
```

Reply:
If it doesn’t do anything, I’d say remove it.

Did you find a new bug introduced by this PR or separately?

The regex you posted looks invalid to me: `(?)` is not a valid part of a regex. But if you are talking about the hash sign inside `[^\\"#]`, here is my analysis:

A hash sign _always_ means a comment inside a heregex (unless it is escaped by a backslash), regardless of _where_ it is. They are allowed in character classes too. So you need to put a backslash in front of the `#` in this case.

Reply:
Sorry, my mistake. This is the correct code. It's from `lexer` (not part of this PR).
```coffeescript
STRING_DOUBLE  =  /// ^(?: [^\\"#] | \\[\s\S] |           \#(?!\{) )* ///
```
I think this bug is "double". It works because of bug you opened #4811 (but it is still a bug).
With this PR the whole part after the hash inside bracets is now consumed.
`#] | \\[\s\S] |           \#(?!\{) )* ///`.
So, hash inside heregex will be treated as a comment unless it is escaped.
I'm fixing this.

Reply:
@lydell I think this is it.
Only unescaped hash is treated as a comment. 
I also corrected some invalid heregex patterns in `lexer.coffee`.
```coffeescript
HEREGEX      = /// ^
  (?:
      [^\\/#]   |
      \\[\s\S]  |
      /(?!//)   |
      \\\#      | # Escaped hash.
      # The comment should consume everything until the end of the line, including the '///'.
      \#(?!\{)[^\n]*\n
  )*
///
```